File: test/scripts/attach_test.py
#
# This file is part of libdebug Python library (https://github.com/libdebug/libdebug).
# Copyright (c) 2023-2024 Gabriele Digregorio, Roberto Alessandro Bertolini. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project root for details.
#

import logging
import unittest

from pwn import process

from libdebug import debugger

logging.getLogger("pwnlib").setLevel(logging.ERROR)


class AttachTest(unittest.TestCase):
    def setUp(self):
        pass

    def test_attach(self):
        r = process("binaries/attach_test")

        d = debugger("binaries/attach_test")
        d.attach(r.pid)
        bp = d.breakpoint("printName", hardware=True)
        d.cont()

        r.recvuntil(b"name:")
        r.sendline(b"Io_no")

        self.assertTrue(d.rip == bp.address)

        d.cont()

        d.kill()


if __name__ == "__main__":
    unittest.main()

File: test/srcs/attach_test.c
//
// This file is part of libdebug Python library (https://github.com/libdebug/libdebug).
// Copyright (c) 2023-2024 Gabriele Digregorio, Roberto Alessandro Bertolini. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for details.
//

#include <stdio.h>

void printName(char name[]) {
    printf("Your name is: %s", name);
}

int main(){
    char name[100];

    setvbuf(stdout, NULL, _IONBF, 0);
    setvbuf(stdin, NULL, _IONBF, 0);

    printf("Enter your name: ");
    fgets(name, sizeof(name), stdin);

    printName(name);
    return 0;
}